opt: fix bug when WITH was used inside WITH RECURSIVE

RaduBerinde opened the pull request:
We were incorrectly building the inner CTE above the recursive CTE.

Fixes #45869.

Release note (bug fix): fixed query errors in cases where a CTE is used inside a
recursive CTE.

Release justification: Category 2: Bug fixes and low-risk updates to new functionality.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/45877)
<!-- Reviewable:end -->

RaduBerinde committed:
opt: fix bug when WITH was used inside WITH RECURSIVE

We were incorrectly building the inner CTE above the recursive CTE.

Fixes #45869.

Release note (bug fix): fixed query errors in cases where a CTE is used inside a
recursive CTE.

Release justification: Category 2 (bug fixes and low-risk updates to new functionality).

andy-kimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45877#-:-M26sjTk1-s3w3UgJtUk:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball)



<!-- Sent from Reviewable.io -->

RaduBerinde commented:
bors r+

RaduBerinde commented:
bors r+
